pennylane/qnode_new/qnode.py

```diff
@@ -0,0 +1,935 @@
+
+
+    def _check_circuit(self, res):
+        """Check that the generated operation queue corresponds to a valid quantum circuit.
+
+        .. note:: The validity of individual Operations is checked already in :meth:`_append_op`.
+
+        Args:
+            res (Sequence[Observable], Observable): output returned by the quantum function
+
+        Raises:
+            QuantumFunctionError: an error was discovered in the circuit
+        """
+        # check the return value
+        if isinstance(res, plo.Observable):
+            if res.return_type is plo.Sample:
+                # Squeezing ensures that there is only one array of values returned
+                # when only a single-mode sample is requested
+                self.output_conversion = np.squeeze
+            else:
+                self.output_conversion = float
+            self.output_dim = 1
+            res = (res,)
+        elif isinstance(res, Sequence) and res and all(isinstance(x, plo.Observable) for x in res):
+            # for multiple observables values, any valid Python sequence of observables
+            # (i.e., lists, tuples, etc) are supported in the QNode return statement.
+
+            # Device already returns the correct numpy array, so no further conversion is required
+            self.output_conversion = np.asarray
+            self.output_dim = len(res)
+            res = tuple(res)
+        else:
+            raise QuantumFunctionError("A quantum function must return either a single measured observable "
+                                       "or a nonempty sequence of measured observables.")
+
+        # check that all returned observables have a return_type specified
+        for x in res:
+            if x.return_type is None:
+                raise QuantumFunctionError(
+                    "Observable '{}' does not have the measurement type specified.".format(x))
+
+        # check that all ev's are returned, in the correct order
+        if res != tuple(self.obs_queue):
+            raise QuantumFunctionError(
+                "All measured observables must be returned in the order they are measured.")
+
+        # check that no wires are measured more than once
+        m_wires = list(w for ob in res for w in ob.wires)
+        if len(m_wires) != len(set(m_wires)):
+            raise QuantumFunctionError('Each wire in the quantum circuit can only be measured once.')
+
+        self.ops = self.queue + list(res)
+        del self.queue
+        del self.obs_queue
+
+        # True if op is a CV, False if it is a discrete variable (Identity could be either)
+        are_cvs = [isinstance(op, plo.CV) for op in self.ops if not isinstance(op, plops.Identity)]
+        if not all(are_cvs) and any(are_cvs):
+            raise QuantumFunctionError(
+                "Continuous and discrete operations are not allowed in the same quantum circuit.")
```

Comment:
Once we use `{model: qubit}` etc. in the Device.capabilities, this can be deleted :)

That opens up the option in the future for a device that supports both CV and qubit models.

Reply:
I'm actually really looking forward to such a device, that will open up some cool possibilities!

Reply:
It requires someone to write a hybrid CV-qubit simulator first :laughing: Oddly enough, we have had requests in the past for one.

Reply:
How do you couple qubits and qumodes? You'd need to create a whole new set of gates for this.